Sources/ApodiniExtension/HTTPInformation+LoggingMetadata.swift

```diff
@@ -0,0 +1,74 @@
+//
+// This source file is part of the Apodini open source project
+//
+// SPDX-FileCopyrightText: 2019-2021 Paul Schmiedmayer and the Apodini project authors (see CONTRIBUTORS.md) <paul.schmiedmayer@tum.de>
+//
+// SPDX-License-Identifier: MIT
+//    
+
+import Foundation
+import Apodini
+import Logging
+
+public struct LoggingMetadataInformation: Information {
```

Comment:
Could you please move this file out of ApodiniExtension into ApodiniObserve? It is purely related to logging and not part of any ApodiniExtension APIs.

Reply:
Well the problem is that then `ApodiniGRPC`, `ApodiniWebSocket` etc. would all need to depend on `ApodiniObserve`. That's not the goal of `ApodiniObserve`, maybe we will even break it out to a separate repo sometime.

Since I didn't want to place the file in `Apodini` itself, i did it in `ApodiniExtension`. Do you have other suggestions? :)

Reply:
I'd suggest to not think in terms of use-cases but in terms of features:
* introduce a new protocol `StringKeyedEncodableInformation` (in ApodiniExtension), which requires a string key plus an encodable value
* introduce a new struct `TextInformation` (in ApodiniExtension), which just has a string key and string value
* conform existing HTTPInformation types and `TextInformation` to `StringKeyedEncodableInformation` (in ApodiniExtension)
* move `LoggingMetadataInformation` to ApodiniObserve and make it initializable from a `StringKeyedEncodableInformation`

Would that work?

Reply:
Yep that should work, I'll take a closer look in the evening

Reply:
Instead of creating weird generic defaults in ApodiniExtension which only get used for a very specific use case, can we do the following:

* Move the `LoggingMetadataInformation` into `ApodiniObserve`. It is specific to this package in is only every used exactly by this package.
* **don't** create an explicit dependency to `ApodiniObserve ` in the Rest and gRPC exporters but use `canImport(ApodiniObserve)` in those exporters to add the `LoggingMetadataInformation` to the InformationSet.

In the example of gRPC one could rewrite the `getLoggingMetadataInformation` function like the following: 

```swift
 static func getLoggingMetadataInformation(_ message: GRPCMessage) -> [AnyInformation] {
     #if canImport(ApodiniObserve) 
         [
             LoggingMetadataInformation(key: .init("data"), metadataValue: message.data.count <= 32_768 ? .string(message.data.base64EncodedString()) : .string("\(message.data.base64EncodedString().prefix(32_715))... (Further bytes omitted since data too large!)")),
             LoggingMetadataInformation(key: .init("dataLength"), metadataValue: .string(message.length.description)),
             LoggingMetadataInformation(key: .init("compression"), metadataValue: .string(message.compressed.description)),
             LoggingMetadataInformation(key: .init("didCollectAllFragments"), metadataValue: .string(message.didCollectAllFragments.description))
         ]
     #else
         []
     #endif
 }
```

At least this is how I expect `canImport` to work, right? (help @theMomax).

If this shouldn't be possible, I would advocate for creating a separate `ApodiniLoggingSupport` package (which could potentially be reused by other Logging facility than ApodiniObserve) which just contains the Information definitions.

Reply:
I also thought about such approach @Supereg and I think that should work as you described. However, I think for the `Information` API to be successful in the long run we have to have common "weird" interfaces that new information can conform to and automatically gain compatibility with most features.

If I were to implement a new `Information`, I'd much prefer a "Just conform to either `StringKeyedEncodableInformation` or <insert-other-future-for-example-integer-based-protocol> and you should be fine" strategy over a "Here is a list of 10 common use-cases you basically want to support with your `Information` and they all define custom protocols you have to conform to, so have fun implementing all of them".

Reply:
> If I were to implement a new Information, I'd much prefer a "Just conform to either StringKeyedEncodableInformation or and you should be fine" strategy over a "Here is a list of 10 common use-cases you basically want to support with your Information and they all define custom protocols you have to conform to, so have fun implementing all of them".

I agree with you in the general case. I think I proposed a similar approach in https://github.com/Apodini/Apodini/pull/330#discussion_r685154826 building on top of the InformationClass feature.
In any case, such a feature should be built around the `InformationClass` feature, providing some default and generic InformationClasses in the main Apodini package.

But I don't think this applies to this case here. `LoggingMetadataInformation` is a specific Information with a specific use case which is not just a string key-value pair but has dedicated **semantics**. While it might conform to a `StringKeyedCodableInformationClass` or to a more specific `StringKeyedStringInformationClass` the information itself can't be a generic `TextInformation`. It transports logging information, and I my opinion this should be made explicit.

In every case, the Information instance should always be specific to the use case. However `InformationClasses` can be used to facilitate reuse and genericness.
E.g. the `AnyHTTPInformation` could also just be a `TextInformation` as its simply a string key-value-pair. But a `AnyHTTPInformation` carries important context about the origin and use of that information. Its pretty much the same think, in my view.

Whats your opinion on that?

Reply:
I think I'm not too far from your position on this. The concern I have with use-case specific `Information` or `InformationClass`es is that the universe of use-cases has infinite dimensions. As far as I understood your concern with semantics is that e.g. a password HTTP header (which would conform to `StringKeyedCodableInformationClass`) should not be logged. And I totally agree. However, once we make the `InterfaceExporter` responsible for this differentiation, the number of decisions and implementations that have to be made explodes. And I fear that this would end up with most exporter - use case pairs to just have no integration whatsoever.

Maybe we should pursue an attribute-based approach. I.e. the `StringKeyedCodableInformationClass` has requires a set of attributes to be defined by the exporter. The attributes that exist are defined in ApodiniExtension and use either neutral or the exporters' language. E.g. instead of having attributes "suitableForLogging", "requiredForCacheEquality" and "requiredForAuthentication" , we'd have "containsCredentials". What do you think @Supereg ?

Reply:
I think we are pretty much on the same page, but talk past each other.

Maybe to summarize my views.

* Introduce a `StringKeyedCodableInformationClass` or a `StringKeyedStringInformationClass` protocol (maybe the `StringKeyedStringInformationClass` could even just be a more specializes version of the codable one using protocol inheritance). The only difference to your original idea is that it builds upon the `InformationClass` protocol/feature, giving the whole thing a name.
The `LoggingMetadataInformation` would be part of the `StringKeyedCodableInformationClass`.
The existing `AnyHTTPInformation` would also be part of the `StringKeyedStringInformationClass` (again, the `HTTPHeaderInformationClass` protocol might just inherit from the `StringKeyedStringInformationClass`)!
* I'm opposed to the idea of introducing a generic `TextInformation` (from your original comment). What information does it transport? Where should the information be used? I think any `Information` **instance** should always be use-case-specific (e.g. `LoggingMetadataInformation` or `AnyHTTPInformation`). We should solely rely on `InformationClass`es for the generalization (by providing some standard ones in the Apodini[Extension] target, like the ones from the previous point). That was my main point from the previous comment.
* I like the idea of marking certain Information to contain sensitive data. I haven't thought about that yet. But particularly a HTTPInformation might transport sensitive information (e.g. the Authorization header). Adding an optional property to the `InformationClass` protocol probably makes the most sense(?)  
Do we have other areas where sensitive data might accidentally be logged?
* `LoggingMetadataInformation` should either be moved into the `ApodiniObserve` target and use `canImport(...)` in the exporters or move it into a separate target which can be imported from the exporters (resembles https://github.com/Apodini/Apodini/pull/330#discussion_r685161204). This again resembles my point of view, that a `Information` **instance** is always create with a predefined context. Therefore, creation of such instances can't be done in a generic way. Only querying can be (InformationClasses).

Reply:
I think the problem I see is that the logger might want to log things that are e.g. normal HTTP headers. They are not explicitly there to be logged, but could be used in a different context too. With your setup, I guess either the logger just queries for LoggingMetadataInformation, or it queries for the broader StringKeyedCodableInformationClass and gets both types (logging and HTTP headers)?

Reply:
> I think the problem I see is that the logger might want to log things that are e.g. normal HTTP headers. They are not explicitly there to be logged, but could be used in a different context too. With your setup, I guess either the logger just queries for LoggingMetadataInformation, or it queries for the broader StringKeyedCodableInformationClass and gets both types (logging and HTTP headers)?

I'm not sure what your point is exactly? With my setup you can query for any `StringKeyedCodableInformationClass`, you can query for only the `HTTPHeaderInformationClass`, there should maybe be also a `LoggingMetadataInformationClass` (or just query for `LoggingMetadataInformation`), and of course you can do any sort of filtering by combining those: querying for `StringKeyedCodableInformationClass` but excluding `HTTPHeaderInformationClass` to get any key-value style Information instances which are not HTTP headers.

Reply:
Ok, if that is possible, that is probably the best solution we have right now 👍🏻. You're the specialist on that topic😁

Reply:
@Supereg Please take a look at how I implemented it.
I had to introduce a separate target `ApodiniLoggingSupport` as discussed, since I couldn't get `canImport(...)` to work (no clue why, I got ever-changing error messages). The target contains all "log helpers", which would allow for easy extensibility.
 I'm not quite sure how to express the inheritance from `StringKeyedCodableInformationClass ` to `StringKeyedStringInformationClass` without associated types etc. Could you give me some pointers here?
I also made the `Authorization` header type `sensitive`, but the default of the `InformationClass` is false (since probably most information headers don't contain sensitive data).